src/hotspot/share/classfile/modules.cpp

```diff
@@ -611,54 +637,28 @@ void Modules::dump_addmods_names() {
   if (addmods_names != nullptr) {
     _archived_addmods_names = ArchiveBuilder::current()->ro_strdup(addmods_names);
   }
-  ArchivePtrMarker::mark_pointer(&_archived_addmods_names);
+}
+
+const char* Modules::get_addmods_names_as_sorted_string() {
+  return get_numbered_property_as_sorted_string("jdk.module.addmods", Arguments::addmods_count());
 }
 
 void Modules::serialize_addmods_names(SerializeClosure* soc) {
   soc->do_ptr(&_archived_addmods_names);
   if (soc->reading()) {
-    bool disable = false;
-    if (_archived_addmods_names[0] != '\0') {
-      if (Arguments::addmods_count() == 0) {
-        log_info(cds)("--add-modules module name(s) found in archive but not specified during runtime: %s",
-            _archived_addmods_names);
-        disable = true;
-      } else {
-        const char* addmods_names = get_addmods_names_as_sorted_string();
-        if (strcmp((const char*)_archived_addmods_names, addmods_names) != 0) {
-          log_info(cds)("Mismatched --add-modules module name(s).");
-          log_info(cds)("  dump time: %s runtime: %s", _archived_addmods_names, addmods_names);
-          disable = true;
-        }
-      }
-    } else {
-      if (Arguments::addmods_count() > 0) {
-        log_info(cds)("--add-modules module name(s) specified during runtime but not found in archive: %s",
-                      get_addmods_names_as_sorted_string());
-        disable = true;
-      }
-    }
-    if (disable) {
-      log_info(cds)("Disabling optimized module handling");
-      CDSConfig::stop_using_optimized_module_handling();
-    }
-    log_info(cds)("optimized module handling: %s", CDSConfig::is_using_optimized_module_handling() ? "enabled" : "disabled");
-    log_info(cds)("full module graph: %s", CDSConfig::is_using_full_module_graph() ? "enabled" : "disabled");
-
-    // Don't hold onto the pointer, in case we might decide to unmap the archive.
-    _archived_addmods_names = nullptr;
+    check_archived_flag_consistency(&_archived_addmods_names, get_addmods_names_as_sorted_string(), "jdk.module.addmods");
   }
 }
 
-const char* Modules::get_addmods_names_as_sorted_string() {
+const char* Modules::get_numbered_property_as_sorted_string(const char* property, unsigned int property_count) {
   ResourceMark rm;
   const int max_digits = 3;
   const int extra_symbols_count = 2; // includes '.', '\0'
-  size_t prop_len = strlen("jdk.module.addmods") + max_digits + extra_symbols_count;
+  size_t prop_len = strlen(property) + max_digits + extra_symbols_count;
   char* prop_name = resource_allocate_bytes(prop_len);
   GrowableArray<const char*> list;
-  for (unsigned int i = 0; i < Arguments::addmods_count(); i++) {
-    jio_snprintf(prop_name, prop_len, "jdk.module.addmods.%d", i);
+  for (unsigned int i = 0; i < property_count; i++) {
```

Comment:
There's no need for property_count. You can just keep reading until Arguments::get_property() returns null. That way, there's no need to expose `Arguments::addmods_count()`, etc.

```
for (int i = 0; ; i++) {
  jio_snprintf(prop_name, prop_len, "%s.%d", property, i);
  const char* prop_value = Arguments::get_property(prop_name);
  if (prop_value == nullptr) {
    break;
  }
  ...
}
```

Also, for existing code: it's better to use a few extra bytes so you never need to worry about buffer overflow.

```
// theoretical string size limit for decimal int, but the following loop will end much sooner due to
// OS command-line size limit.
const int max_digits = 10;
```